zebra/zapi_msg.c

```diff
@@ -1586,8 +1586,45 @@ static void zread_route_add(ZAPI_HANDLER_ARGS)
 	if (CHECK_FLAG(api.message, ZAPI_MESSAGE_SRCPFX))
 		src_p = &api.src_prefix;
 
+
+	/**
+	 * zebra_nhg cached handling to speed up route installs
+	 * for routes being sent with the same nexthops in a row.
+	 */
+	assert(sizeof(api.nexthops) == sizeof(client->inhg_cache.nexthops));
+
+	/* If matches cached nexthops, use that NHE ID now */
+	if (memcmp(api.nexthops, client->inhg_cache.nexthops,
+		   sizeof(client->inhg_cache.nexthops))
+	    == 0) {
+		if (IS_ZEBRA_DEBUG_RECV)
+			zlog_debug("%s: Using cached zebra_nhg", __func__);
+
+		/* Just pass down with ID */
+		nexthop_group_delete(&re->ng);
+		re->nhe_id = client->inhg_cache.nhe_id;
+	} else {
+		/**
+		 * De-ref the old NHE.
+		 * Reset the cached ID to 0 so we know to
+		 * get the next created NHE and use it.
+		 */
+		zebra_nhg_decrement_ref(
+			zebra_nhg_lookup_id(client->inhg_cache.nhe_id));
+		client->inhg_cache.nhe_id = 0;
```

Comment:
Does the hash algo - "jhash", I guess - actually prevent zero from being computed as a valid value?

Reply:
zero is not allowed in the kernel.  Zero to the kernel means you want the kernel to assign and ID. I pretty much follow the same semantic in `rib_add_multipath()` if its not zero, we should be doing a lookup and the `route_entry` should use that NHE. If its zero, we hash the whole group to find/create a new entry with id != 0.

The setting of the `cache.nhe_id` to 0 here though is to indicate we have cleared the cache and it needs to be re-cached after the next rib_add_multipath() call.

I considered using a `cleared` boolean in addition to make it more clear. Would that be better?